cou/apps/core.py

```diff
@@ -106,7 +106,44 @@ def post_upgrade_steps(
         :return: List of post upgrade steps.
         :rtype: list[PostUpgradeStep]
         """
-        return self._get_enable_scheduler_step(units) + super().post_upgrade_steps(target, units)
+        return (
+            self._get_enable_scheduler_step(units)
+            + self._get_restart_subordinate_services_steps(units)
+            + super().post_upgrade_steps(target, units)
+        )
+
+    def _get_restart_subordinate_services_steps(
+        self, units: Optional[list[Unit]]
+    ) -> list[PostUpgradeStep]:
+        """Get step to restart all subordinate services if they aren't running.
+
+        :param units: Units to restart subordinate services.
+        :type units: Optional[list[Unit]]
+        :return: Steps to restart to subordinate services
+        :rtype: list[PostUpgradeStep]
+        """
+        if not units:
+            return []
+        steps = []
+        for unit in units:
+            for subordinate in unit.subordinates:
+                if "ceilometer-agent" in subordinate.name:
```

Comment:
relying on application name is not robust. Unfortunately it seems that there is no link on python libjuju between the subordinate unit and the actual charm name. I tried using juju 3.1 and the charm field is empty:

```shell
>>> status.applications["ubuntu"].units["ubuntu/0"].subordinates["hardware-observer/0"].charm
''
```

Reply:
Changed to charm_name on newest commit.